### PR TITLE
Added comments to BaseConfig saving YAML

### PR DESCRIPTION
**Pull request by wsp-mbuckley**
Describe Changes
---
Added optional datetime and custom comments to the top of YAML config files.
Added optional comments to the write example method.

Address #57 

Task Checklist
----
- [x] Have unittests been added (testing individual functions and their edge cases)
- [ ] ~~Have integration tests been added (if applicable, to test modules of functionality)~~
- [x] Updated RELEASE.md to reflect changes in PR
- [ ] ~~Have new dependencies been added?~~
  - [ ] ~~Have they been added to either `requirements.txt` or `requirements_dev.txt`.~~
  - [ ] ~~Have they been added to `pyproject.toml`~~

### CODE REVIEW

**Comment by wsp-mbuckley** [comment]
@isaac-tfn this will add date / time comments when saving to YAML by default and provides another parameter for custom comments if required, let me know what you think

File: src/caf/toolkit/config_base.py [hunk 1]
```diff
@@ -132,27 +133,68 @@ def to_yaml(self) -> str:
 
         return strictyaml.as_document(json_dict).as_yaml()
 
-    def save_yaml(self, path: Path) -> None:
+    def save_yaml(
+        self,
+        path: Path,
+        datetime_comment: bool = True,
+        other_comment: str | None = None,
```
Comment (isaac-tfn) [inline src/caf/toolkit/config_base.py:140]: Sorry to do this every time, but this isn't compatible with 3.9

File: tests/test_config_base.py [hunk 2]
```diff
@@ -287,7 +309,84 @@ def test_example(self, path: pathlib.Path) -> None:
...
+                "# already formatted comment with newlines",
+            ),
+        ],
+    )
+    @pytest.mark.parametrize("format_", [True, False])
+    def test_other_comment(
+        self,
+        basic: ConfigTestClass,
+        tmp_path: pathlib.Path,
+        comment: str | None,
```
Comment (isaac-tfn) [inline tests/test_config_base.py:361]: Same as my other comment just commenting again to find it easily

**Review by isaac-tfn — COMMENTED** [review]
Everything looks good apart from my comments on the typing not being backwards compatible.

**Comment by wsp-mbuckley** [comment]
> Everything looks good apart from my comments on the typing not being backwards compatible.

Thanks for having a look, I've sorted that now so it should be ready to merge

**Comment by isaac-tfn** [comment]
There's a test failing which hasn't been touched in this PR (test_array_utils.TestSparseSum). Can you spot a reason why @BenTaylor-TfN?

**Comment by isaac-tfn** [comment]
> There's a test failing which hasn't been touched in this PR (test_array_utils.TestSparseSum). Can you spot a reason why @BenTaylor-TfN?

I've just spotted there is a comment in the test that this is a known bug

**Comment by wsp-mbuckley** [comment]
> > There's a test failing which hasn't been touched in this PR (test_array_utils.TestSparseSum). Can you spot a reason why @BenTaylor-TfN?
> 
> I've just spotted there is a comment in the test that this is a known bug

Is this a linux specific issue because the test is passing fine on my machine

**Comment by BenTaylor-TfN** [comment]
Just spotted that this has also just happened on main, with some code that was previously fine before the PR was merged. See here:
https://github.com/Transport-for-the-North/caf.toolkit/actions/runs/5953829958/job/16148999342

Here, it happened in python 3.11 and windows, must be something random causing it. Will try narrow it down shortly.
Once I've logged this error, I'll re-run the test and it will likely pass.

**Comment by BenTaylor-TfN** [comment]
@wsp-mbuckley , sorted now

**Comment by wsp-mbuckley** [comment]
@BenTaylor-TfN or @isaac-tfn I think both of you are happy with this, please could one of you approve it and I'll merge it into main 😀

**Review by isaac-tfn — APPROVED** [review]

**Pull request merged**